#  Hershberger et al. Transcriptome-wide association and prediction for carotenoids and tocochromanols in fresh sweet corn kernels
![language: R](https://img.shields.io/badge/language-R-blue.svg)

# PREPRINT

This repository and website documents all analyses, summary, tables and figures associated with the following PREPRINT: [Transcriptome-wide association and prediction for carotenoids and tocochromanols in fresh sweet corn kernels](https://doi.org/10.1101/2021.09.24.461734)!

# Abstract

Sweet corn is consistently one of the most highly consumed vegetables in the U.S., providing a valuable opportunity to increase nutrient intake through biofortification. Significant variation for carotenoid (provitamin A, lutein, zeaxanthin) and tocochromanol (vitamin E, antioxidants) levels is present in temperate sweet corn germplasm, yet previous genome-wide association studies (GWAS) of these traits have been limited by low statistical power and mapping resolution. Here, we employed a high-quality transcriptomic dataset collected from fresh sweet corn kernels to conduct transcriptome-wide association studies (TWAS) and transcriptome prediction studies for 39 carotenoid and tocochromanol traits. In agreement with previous GWAS findings, TWAS detected significant associations for four causal genes, _β-carotene hydroxylase_ (_crtRB1_), _lycopene epsilon cyclase_ (_lcyE_), _γ-tocopherol methyltransferase_ (_vte4_), and _homogentisate geranylgeranyltransferase_ (_hggt1_) on a transcriptome-wide level. Pathway-level analysis revealed additional associations for _deoxy-xylulose synthase2_ (_dxs2_), _diphosphocytidyl methyl erythritol synthase2_ (_dmes2_), _cytidine methyl kinase1_ (_cmk1_), and _geranylgeranyl hydrogenase1_ (_ggh1_), of which, _dmes2_, _cmk1_, and _ggh1_ have not previously been identified through maize association studies. Evaluation of prediction models incorporating genome-wide markers and transcriptome-wide abundances revealed a trait-dependent benefit to the inclusion of both genomic and transcriptomic data over solely genomic data, but both transcriptome- and genome-wide datasets outperformed a priori candidate gene-targeted prediction models for most traits. Altogether, this study represents an important step towards understanding the role of regulatory variation in the accumulation of vitamins in fresh sweet corn kernels.

# Data availability
Much of the supporting data and output from the analyses documented here are too large for GitHub.

The `data` and `output` folders for this repository will be mirrored [here](https://datacommons.cyverse.org/browse/iplant/home/shared/GoreLab/dataFromPubs/Hershberger_SweetCornRNA_2021).
